websock/session.nim

```diff
@@ -347,8 +352,8 @@ proc recv*(
         trace "Reading bytes from frame stream", len
         let read = await ws.frame.read(ws.stream.reader, addr pbuffer[consumed], len)
         if read <= 0:
-          trace "Didn't read any bytes, breaking"
-          break
+          trace "Didn't read any bytes, stopping"
+          raise newException(WSClosedError, "WebSocket is closed!")
```

Comment:
read = 0 means EOF

Reply:
the point with returning 0 on EOF is that exceptions are generally not suitable for entirely normal behavior - ie this way, a 100% correct application that closes the stream regularly cannot work without at least one exception being fired - I'd generally expect this of any `recv`-style function (recv itself in bsd sockets works this way - it returns 0 on EOF and -1 on errors)

Reply:
(to be clear, this `recv` function should also return 0 on a regular stream closure)

Reply:
Like often, the line between "regular behavior" and "exceptional behavior" gets fuzzy
Here for instance, this is not a regular behavior, because the client is closing the connection "unsafely", in the middle of the frame

More importantly, websocket allows empty frame, which would "legally" return 0 from `recvMsg`
So unfortunately, from this procedure, the only way I currently have to tell `recvMsg` that the stream is closed is through an exception

Reply:
> Here for instance, this is not a regular behavior, because the client is closing the connection "unsafely", in the middle of the frame

ah, fair point then - in general cases, it's not much of a "fuzzy" border for `recv` - it's more of a convention that's quite common (which `readOnce` generally continues, as a tradition), that closing is not considered an error at all (regardless of error reporting method) if it's done through a "regular" `close`